### ISSUE TITLE
Memory leak leads to heap overflow

### ISSUE THREAD

**Issue by pkovac** [reactiflux/discord-irc#330]
14 Oct 15:06:57 - Received: :nova.esper.net PONG nova.esper.net :115367

<--- Last few GCs --->

[2539:0x4299a10] 1791041298 ms: Mark-sweep 1325.7 (1350.2) -> 1323.7 (1350.2) MB, 718.2 / 0.1 ms  allocation failure GC in old space requested
[2539:0x4299a10] 1791042276 ms: Mark-sweep 1323.7 (1350.2) -> 1323.6 (1350.2) MB, 977.4 / 0.0 ms  allocation failure GC in old space requested
[2539:0x4299a10] 1791043255 ms: Mark-sweep 1323.6 (1350.2) -> 1323.6 (1350.2) MB, 979.5 / 0.0 ms  last resort 
[2539:0x4299a10] 1791044225 ms: Mark-sweep 1323.6 (1350.2) -> 1323.3 (1350.2) MB, 969.6 / 0.0 ms  last resort 

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2a90f6a29891 <JS Object>
    2: heartbeat [/home/sfnet-discord-bot/node-v8.1.2-linux-x64/lib/node_modules/discord-irc/node_modules/discord.js/src/client/websocket/WebSocketConnection.js:~407] [pc=0x1141bd0b5fab](this=0x3bc473b2d581 <an EventEmitter with map 0x263359c3e3d1>,time=0x222777302311 <undefined>)
    3: arguments adaptor frame: 0->1
    4: _onTimeout [/home/sfnet-discord-bot/node-v8.1.2-linux-x64/lib/node_modu...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [node]
 2: 0x13647ec [node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
 5: v8::internal::Factory::NewUninitializedFixedArray(int) [node]
 6: 0xe90ca3 [node]
 7: v8::internal::Runtime_GrowArrayElements(int, v8::internal::Object**, v8::internal::Isolate*) [node]
 8: 0x1141bc30437d
Aborted

**Comment by Throne3d** [comment]
I don't understand much of what's written here. From the stack trace, it looks like… the very last… call… was from `discord.js`? I don't think that helps track the problem down, though.

Do you have any more information on the problem? I've never noticed it, so I wonder if you're doing something weird in your setup (are you on a low-memory machine?).

**Comment by pkovac** [comment]
No, I was running into the Node v8 default heap limit. There's not enough info here, but you can probably find it fairly quickly if you try pulling in heapdump from npm and take some snapshots as the bot is running:
https://blog.risingstack.com/finding-a-memory-leak-in-node-js/

I'll look into it a bit more at some point, but my fix for now was to just stick discord-irc into a systemd unit so it autorestarts when this happens.

**Comment by sethlesky** [comment]
Have there been any updates on this?

I'm having what appears to be a similar issue as seen in the following log:

```
error Command failed with exit code 134.
 
Aborted 
10: 0x15a8c4bc3b67 
9: 0x8e62c6 [/nodejs/bin/node] 
8: node::StringBytes::Encode(v8::Isolate*, char const*, unsigned long, node::encoding, v8::Local<v8::Value>*) [/nodejs/bin/node] 
7: v8::String::NewFromUtf8(v8::Isolate*, char const*, v8::NewStringType, int) [/nodejs/bin/node] 
6: v8::internal::Factory::NewStringFromUtf8(v8::internal::Vector<char const>, v8::internal::PretenureFlag) [/nodejs/bin/node] 
5: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/nodejs/bin/node]
4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/nodejs/bin/node]
3: v8::Utils::ReportOOMFailure(char const*, bool) [/nodejs/bin/node]
2: 0x8ccf9c [/nodejs/bin/node]
1: node::Abort() [/nodejs/bin/node]
 
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 
3: unpack [/app/node_modules/discord.js/src/client/websocket/WebSocketConnection.js:~170] [pc=0x15a8c49070bd](this=0x29ae8a484649 <EventEmitter map =…
2: arguments adaptor frame: 0->3
1: toString [buffer.js:~609] [pc=0x15a8c495b041](this=0x253f2b47cc99 <Uint8Array map = 0x32dd30a43a71>,encoding=0x282dd32822d1 <undefined>,start=0x282dd32822d1 <undefined>,end=0x282dd32822d1 <undefined>)
 
Security context: 0x2af5574a58b9 <JSObject>

==== JS stack trace =========================================
 
<— JS stacktrace —>
[27:0x3001070] 70137123 ms: Mark-sweep 1292.3 (1470.9) -> 1292.3 (1470.4) MB, 2666.6 / 0.0 ms  last resort GC in old space requested
 
[27:0x3001070] 70134456 ms: Mark-sweep 1292.3 (1511.9) -> 1292.3 (1470.9) MB, 2829.5 / 0.0 ms  last resort GC in old space requested
 
[27:0x3001070] 70131626 ms: Mark-sweep 1292.4 (1509.9) -> 1292.3 (1511.9) MB, 2763.2 / 0.0 ms  allocation failure GC in old space requested
 
 
<— Last few GCs —>
```

**Comment by ekmartin** [comment]
Looking through the discord.js issues related to memory leaks [0] it seems like they cache a lot of things without actually clearing the cache. It doesn't seem like this is something they intend to fix: https://github.com/discordjs/discord.js/issues/1409#issuecomment-433648959

It's on the other hand possible to disable some of these caches [1] so maybe that's what we want to do?

[0] https://github.com/discordjs/discord.js/search?q=memory+leak&type=Issues
[1] https://github.com/discordjs/discord.js/pull/2883

**Comment by sethlesky** [comment]
@ekmartin Have you had any luck with disabling the cache?

It looks like the fork was abandoned due to unexpected behavior when disabling stores. https://github.com/discordjs/discord.js/pull/2883#issuecomment-433649130

Our Discord Bot's heap can grow to 2gigs in a couple of hours resulting in a crash. And we have under 50 Discord groups using the bot, so this type of rapid memory expansion seems odd, given there are some bots with pretty massive usage.

Thanks for pointing me to those threads. 👍

**Comment by Throne3d** [comment]
Is this still an issue? I'm looking into updating discord.js, but I don't know if they've made any improvements here.
